File: leetcode/188.买卖股票的最佳时机 IV.py
class Solution:
    def maxProfit(self, k: int, prices: List[int]) -> int:
        n = len(prices)
        @cache
        def dfs(i: int, j: int, hold: bool) -> int:
            if j < 0:
                return -inf
            if i < 0:
                return -inf if hold else 0
            if hold:
                return max(dfs(i - 1, j, True), dfs(i - 1, j - 1, False) - prices[i])
            return max(dfs(i - 1, j, False), dfs(i - 1, j, True) + prices[i])
        return dfs(n - 1, k, False)
    def maxProfit(self, k: int, prices: List[int]) -> int:
        n = len(prices)
        # f[i][k][0] = max(f[i-1][k][0], f[i-1][k-1][1] + prices[i])
        # f[i][k][1] = max(f[i-1][k][1], f[i-1][k-1][0] - prices[i])
        # f[i+1][j+1][0] = max(f[i][j+1][0], f[i][j][1] + prices[i])
        # f[i+1][j+1][1] = max(f[i][j+1][1], f[i][j+1][0] - prices[i])
        f = [[[-inf, -inf] for _ in range(k+2)] for _ in range(n+1)]
        for i in range(1, k+2): f[0][i][0] = 0
        for i in range(n):
            for j in range(k+1):
                f[i+1][j+1][0] = max(f[i][j+1][0], f[i][j+1][1] + prices[i])
                f[i+1][j+1][1] = max(f[i][j+1][1], f[i][j][0] - prices[i])
        return f[n][k+1][0]
    

File: leetcode/188.买卖股票的最佳时机 IV.go
func maxProfit1(k int, prices []int) int {
    n := len(prices)
    memo := make([][][2]int, n)
    for i := range memo {
        memo[i] = make([][2]int, k+1)
        for j := range memo[i] {
            memo[i][j] = [2]int{-1, -1} // -1 表示还没有计算过
        }
    }
    var dfs func(int, int, int) int
    dfs = func(i, j, hold int) (res int) {
        if j < 0 {
            return math.MinInt / 2 // 防止溢出
        }
        if i < 0 {
            if hold == 1 {
                return math.MinInt / 2
            }
            return
        }
        p := &memo[i][j][hold]
        if *p != -1 { // 之前计算过
            return *p
        }
        defer func() { *p = res }() // 记忆化
        if hold == 1 {
            return max(dfs(i-1, j, 1), dfs(i-1, j-1, 0)-prices[i])
        }
        return max(dfs(i-1, j, 0), dfs(i-1, j, 1)+prices[i])
    }
    return dfs(n-1, k, 0)
}
func maxProfit(k int, prices []int) int {
    n := len(prices)
    f := make([][][2]int, n+1)
    for i := range f{
        f[i] = make([][2]int, k+2)
        for j := range f[i]{
            f[i][j] = [2]int{math.MinInt / 2, math.MinInt / 2}
        }
    }
    for j := 1; j <= k+1; j++{
        f[0][j][0] = 0
    }
    for i, p := range prices{
        for j := 1; j <= k+1; j++{
            f[i+1][j][0] = max(f[i][j][0], f[i][j][1] + p)
            f[i+1][j][1] = max(f[i][j][1], f[i][j-1][0] - p)
        }
    }
    return f[n][k+1][0]
}
func max(a, b int) int { if a < b { return b }; return a }